Added route_to_market and lot_route_to_market

trevorcummingsCCS opened the pull request:
SQL Fiddle had  trouble with a primary key against two columns but as postgres  lot_sectors coped with it I am thinking that lot_route_to_market will cope too.

trevorcummingsCCS committed:
Added route_to_market and lot_route_to_market

SQL Fiddle had  trouble with a primary key against two columns but as postgres  lot_sectors coped with it I am thinking that lot_route_to_market will cope too.

adrianmilne commented:
> SQL Fiddle had trouble with a primary key against two columns but as postgres lot_sectors coped with it I am thinking that lot_route_to_market will cope too.

Postgres does have a problem with `lot_sectors` too - the syntax for creating compound keys is not quite correct - it gives this error:

```
ERROR:  multiple primary keys for table "lot_sectors" are not allowed
LINE 56:   sector_code                       VARCHAR(2) PRIMARY KEY
```

The correct syntax is:

```
CREATE TABLE lot_sectors (
  lot_id                            INTEGER,
  sector_code                       VARCHAR(2),
  PRIMARY KEY (lot_id, sector_code)
);
```
I'll update the PR inline with snippets from code that I have working on SBX1.

Also - be good to update the DROP table script too if possible.

adrianmilne commented on agreements/ddl.sql:
```diff
@@ -57,6 +57,25 @@ CREATE TABLE lot_sectors (
+  route_to_market_description       VARCHAR(2000) NOT NULL
+);
+
+CREATE TABLE lot_route_to_market (
```
Cannot have 2 primary keys defined - syntax needs to be:

```
CREATE TABLE lot_route_to_market (
  lot_id                            INTEGER,
  route_to_market_name              VARCHAR(50),
  start_date                        DATE NOT NULL,
  end_date                          DATE NOT NULL,
  location                          VARCHAR(20), -- For example, Regional or National  
  lot_minimum_value                 NUMERIC(18,4),         
  lot_contract_minimum_weeks        SMALLINT, 
  lot_contract_maximum_weeks        SMALLINT,
  lot_procurement_minimum_quantity  SMALLINT,
  lot_procurement_maximum_quantity  SMALLINT,
  PRIMARY KEY (lot_id, route_to_market_name)
);
```

Same for `lot_sectors`:

```
CREATE TABLE lot_sectors (
  lot_id                            INTEGER,
  sector_code                       VARCHAR(2),
  PRIMARY KEY (lot_id, sector_code)
);
```

trevorcummingsCCS replied:
Will make those changes and add the drop table statement.

adrianmilne commented:
Another minor question (not directly related to this Pull Request) - should `lots.lot_number` and `commercial_agreements.commercial_agreement_number` have a unique constraint?

adrianmilne commented:
Sorry - another general question (I'm just building out the entity objects in the Java code so going a bit more in-depth) - should the `INTEGER` type primary keys be `BIGSERIAL PRIMARY KEY` for consistency with Guided Match (? or not)

adrianmilne commented on agreements/ddl.sql:
```diff
@@ -57,6 +57,25 @@ CREATE TABLE lot_sectors (
+  lot_contract_minimum_weeks        SMALLINT, 
+  lot_contract_maximum_weeks        SMALLINT,
+  lot_procurement_minimum_quantity  SMALLINT,
+  lot_procurement_maximum_quantity  SMALLINT
```
I think there is a bit of a disconnect here with Dave's API.

The only field I think we require for Phase 1 is `buyingMethods`, and also potentially `buyingSystemURL` (which I think are missing). These help to inform the display on the results page of a Guided Match journey (text and buttons).

Also in the API - quantity and contract length are a composite of unit + length, not a single value. We could infer units for length (weeks) - but seems like a bit of a mismatch. However - as mentioned I don't think these other fields are essential for Phase 1 (from current understanding).

trevorcummingsCCS replied:
By buying methoid do you mean, MarketPlace, Further Competition, Direct Award and Auction. If so they are the records in route_to_method. I did forget in my haste to add the buying method. Yes when I was creating the table I did have contract length uom but I thourght that was overkill and just rationalised it and was going to talk to you guys today. I can put them them back in.

adrianmilne commented on agreements/ddl.sql:
```diff
@@ -57,6 +57,25 @@ CREATE TABLE lot_sectors (
 CREATE INDEX LOT_SECTORS_IDX1 on LOT_SECTORS (lot_id,sector_code);
 
+CREATE TABLE route_to_market (
+  route_to_market_name              VARCHAR(50) PRIMARY KEY,
```
I guess using natural keys rather than surrogate keys is a conscious decision? (just curious as to pros/cons out of interest)

trevorcummingsCCS replied:
There are only going to be four or five records in here . I thought creating a surrogate system generated key was over the top.

trevorcummingsCCS committed:
Amended Primary Keys and renamed columns

Added the correct definition for composite primary key columns.
Renamed commecial_agreements.description to commercial_agreements.commercial_agreement_description
Put the uom unit_of_measure columns back as they were defined in the api

trevorcummingsCCS committed:
Added the two new tables to the drop file

trevorcummingsCCS committed:
Amended sector_description from sector_descripton

trevorcummingsCCS committed:
Removed the quantity columns from lot_route_to_market

adrianmilne reviewed (approved):
Provisioned on SBX1 without error 👍

trevorcummingsCCS committed:
Removed column.

Removed authorisation email as the model has been changed so that you could have multiple authorisation email addresses as other types as well.
